File: planning/planning_debug_tools/scripts/perception_replayer/perception_reproducer.py
#!/usr/bin/env python3

# Copyright 2023 TIER IV, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import argparse
from collections import deque
import pickle

from geometry_msgs.msg import PoseWithCovarianceStamped
import numpy as np
from perception_replayer_common import PerceptionReplayerCommon
import rclpy
from utils import StopWatch
from utils import create_empty_pointcloud
from utils import translate_objects_coordinate


class PerceptionReproducer(PerceptionReplayerCommon):
    def __init__(self, args):
        self.rosbag_ego_odom_search_radius = (
            args.search_radius
        )  # (m) the range of the ego odom to search,
        self.ego_odom_search_radius = (
            self.rosbag_ego_odom_search_radius
        )  # it may be set by an individual parameter.

        self.speed_threshold = (
            args.speed_threshold
        )  # (m/s) the speed threshold to determine the ego is stopping.

        super().__init__(args, "perception_reproducer")

        self.reproduce_sequence_indices = deque()  # contains ego_odom_idx

        self.last_published_idx = 0
        self.last_published_idx_in_sequence = 0
        self.prev_ego_odom_msg = None
        self.prev_objects_msg = None
        self.prev_traffic_signals_msg = None

        self.stopwatch = StopWatch(self.args.verbose)  # for debug

        # refresh cool down for setting initial pose in psim.
        def refresh_stat():
            self.last_published_idx = 0
            self.last_published_idx_in_sequence = 0
        self.sub_init_pos = self.create_subscription(
            PoseWithCovarianceStamped, "/initialpose", refresh_stat, 1
        )

        # to make some data to accelerate computation
        self.preprocess_data()

        # start main timer callback
        time_diffs = []
        prev_stamp = self.rosbag_ego_odom_data[0][0]
        for stamp, _ in self.rosbag_ego_odom_data[1:]:
            time_diff = (stamp - prev_stamp) / 1e9
            time_diffs.append(time_diff)
            prev_stamp = stamp

        average_ego_odom_interval = sum(time_diffs) / len(time_diffs)
        # slow down the publication speed.
        average_ego_odom_interval *= args.publishing_speed_factor
        self.timer = self.create_timer(average_ego_odom_interval, self.on_timer)

        # kill perception process to avoid a conflict of the perception topics
        self.timer_check_perception_process = self.create_timer(3.0, self.on_timer_kill_perception)

        print("Start timer callback")

    def preprocess_data(self):
        # closest search with numpy data is much faster than usual
        self.rosbag_ego_odom_data_numpy = np.array(
            [
                [data[1].pose.pose.position.x, data[1].pose.pose.position.y]
                for data in self.rosbag_ego_odom_data
            ]
        )

    def on_timer_kill_perception(self):
        self.kill_online_perception_node()

    def on_timer(self):
        if self.args.verbose:
            print("\n-- on_timer start --")
        self.stopwatch.tic("total on_timer")

        timestamp = self.get_clock().now()
        timestamp_msg = timestamp.to_msg()

        if self.args.detected_object:
            pointcloud_msg = create_empty_pointcloud(timestamp_msg)
            self.pointcloud_pub.publish(pointcloud_msg)

        if not self.ego_odom:
            print("No ego odom found.")
            return

        ego_pose = self.ego_odom.pose.pose
        ego_speed = np.sqrt(self.ego_odom.twist.twist.linear.x**2 +
                            self.ego_odom.twist.twist.linear.y**2)

        # Just get the closest odom's objects if the ego is moving (old reproducer).
        if ego_speed > self.speed_threshold or self.prev_ego_odom_msg is None:
            self.update_counter = 0
            ego_odom_indices = [self.find_nearest_ego_odom_index(ego_pose)]
        elif self.update_counter < 50 and len(self.reproduce_sequence_indices) != 0:
            self.update_counter += 1
            ego_odom_indices = []
        else:
            self.update_counter = 0
            # find the nearby ego odom arange the ego.
            self.stopwatch.tic("find_nearby_ego_odom_indies")
            nearby_ego_odom_indies = self.find_nearby_ego_odom_indies(
                [self.ego_odom.pose.pose], self.ego_odom_search_radius
            )
            nearby_ego_odom_indies = [
                self.rosbag_ego_odom_data[idx][1].pose.pose for idx in nearby_ego_odom_indies
            ]
            if len(nearby_ego_odom_indies) == 0:
                nearby_ego_odom_indies = [self.rosbag_ego_odom_data[self.find_nearest_ego_odom_index(ego_pose)][1].pose.pose
                                          ]

            # find a list of ego odom around the nearby_ego_odom_indies.
            ego_odom_indices = self.find_nearby_ego_odom_indies(
                nearby_ego_odom_indies, self.rosbag_ego_odom_search_radius
            )
            self.stopwatch.toc("find_nearby_ego_odom_indies")

            # speed filter for the ego_odom_indices
            def get_speed(idx):
                log_ego_odom = self.rosbag_ego_odom_data[idx][1]
                return np.sqrt(log_ego_odom.twist.twist.linear.x**2 +
                               log_ego_odom.twist.twist.linear.y**2)
            ego_odom_indices = [idx for idx in ego_odom_indices if get_speed(
                idx) < self.speed_threshold]

        # Filter reproduce_sequence with timestamp before self.last_published_idx
        self.stopwatch.tic("sourt reproduce_sequence")
        if len(ego_odom_indices) != 0:
            last_timestamp = self.rosbag_ego_odom_data[self.last_published_idx][0]
            ego_odom_indices = [
                idx for idx in ego_odom_indices if self.rosbag_ego_odom_data[idx][0] > last_timestamp
            ]
            ego_odom_indices = sorted(ego_odom_indices)
            self.reproduce_sequence_indices = deque(ego_odom_indices)
        self.stopwatch.toc("sourt reproduce_sequence")

        # Get data to publish
        repeat_trigger = len(self.reproduce_sequence_indices) == 0
        if not repeat_trigger:
            ego_odom_idx = self.reproduce_sequence_indices.popleft()
            # if self.update_counter % 5 != 0:
            #     self.last_published_idx = ego_odom_idx
            #     return
        else:
            ego_odom_idx = self.last_published_idx
            self.last_published_idx = ego_odom_idx

        # extract messages by the nearest ego odom timestamp
        pose_timestamp, ego_odom_msg = self.rosbag_ego_odom_data[ego_odom_idx]

        self.stopwatch.tic("find_topics_by_timestamp")
        objects_msg, traffic_signals_msg = self.find_topics_by_timestamp(pose_timestamp)
        self.stopwatch.toc("find_topics_by_timestamp")
        if self.args.verbose:
            print("reproduce_sequence_indices: ", list(self.reproduce_sequence_indices)[:20])

        # Transform and publish current data.
        self.stopwatch.tic("transform and publish")

        # ego odom
        self.recorded_ego_pub.publish(ego_odom_msg)

        # objects
        if objects_msg:
            objects_msg.header.stamp = timestamp_msg
            if self.args.detected_object:
                # copy the messages
                self.stopwatch.tic("message deepcopy")
                objects_msg_copied = pickle.loads(
                    pickle.dumps(objects_msg)
                )  # this is x5 faster than deepcopy
                self.stopwatch.toc("message deepcopy")

                log_ego_pose = ego_odom_msg.pose.pose
                translate_objects_coordinate(ego_pose, log_ego_pose, objects_msg_copied)
            else:
                objects_msg_copied = objects_msg
            self.objects_pub.publish(objects_msg_copied)

        # traffic signals
        if traffic_signals_msg:
            traffic_signals_msg.stamp = timestamp_msg
            self.traffic_signals_pub.publish(traffic_signals_msg)
            self.prev_traffic_signals_msg = traffic_signals_msg
        elif self.prev_traffic_signals_msg:
            self.prev_traffic_signals_msg.stamp = timestamp_msg

            self.traffic_signals_pub.publish(self.prev_traffic_signals_msg)
        self.stopwatch.toc("transform and publish")

        self.stopwatch.toc("total on_timer")

    def find_nearest_ego_odom_index(self, ego_pose):
        # nearest search with numpy format is much (~ x100) faster than regular for loop
        self_pose = np.array([ego_pose.position.x, ego_pose.position.y])
        dists_squared = np.sum((self.rosbag_ego_odom_data_numpy - self_pose) ** 2, axis=1)
        nearest_idx = np.argmin(dists_squared)
        return nearest_idx

    def find_nearby_ego_odom_indies(self, ego_poses: list, search_radius: float):
        ego_poses_np = np.array([[pose.position.x, pose.position.y] for pose in ego_poses])
        dists_squared = np.sum(
            (self.rosbag_ego_odom_data_numpy[:, None] - ego_poses_np) ** 2, axis=2
        )
        nearby_indices = np.where(np.any(dists_squared <= search_radius**2, axis=1))[0]

        return nearby_indices


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("-b", "--bag", help="rosbag", default=None)
    parser.add_argument(
        "-d", "--detected-object", help="publish detected object", action="store_true"
    )
    parser.add_argument(
        "-t", "--tracked-object", help="publish tracked object", action="store_true"
    )
    parser.add_argument(
        "-f", "--rosbag-format", help="rosbag data format (default is db3)", default="db3"
    )
    parser.add_argument(
        "-v", "--verbose", help="output debug data", action="store_true", default=False
    )
    parser.add_argument(
        "-r",
        "--search-radius",
        help="the search radius for searching rosbag's ego odom messages around the nearest ego odom pose (default is 1.5 meters), if the search radius is set to 0, it will always publish the closest message, just as the old reproducer did.",
        type=float,
        default=2.0,
    )
    parser.add_argument(
        "--publishing-speed-factor",
        type=float,
        default=1.2,
        help="A factor to slow down the publication speed.",
    )
    parser.add_argument(
        "-s",
        "--speed-threshold",
        type=float,
        default=2.,
        help="The speed threshold to determine whether to publish in time sequence."
    )

    args = parser.parse_args()

    rclpy.init()
    node = PerceptionReproducer(args)

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

File: planning/planning_debug_tools/src/perception_replayer/perception_reproducer.cpp
// Copyright 2025 TIER IV, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "perception_reproducer.hpp"

#include "utils.hpp"

#include <algorithm>
#include <chrono>
#include <cmath>
#include <deque>
#include <limits>
#include <memory>
#include <numeric>
#include <string>
#include <vector>

namespace autoware::planning_debug_tools
{

PerceptionReproducer::PerceptionReproducer(
  const PerceptionReproducerParam & param, const rclcpp::NodeOptions & node_options)
: PerceptionReplayerCommon(param, "perception_reproducer", node_options), param_(param)
{
  RCLCPP_INFO(get_logger(), "Starting PerceptionReproducer initialization");

  ego_odom_search_radius_ = param_.search_radius;

  // subscription for /initialpose to refresh cool down
  sub_init_pos_ = this->create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>(
    "/initialpose", 1,
    std::bind(&PerceptionReproducer::on_initialpose, this, std::placeholders::_1));

  // calculate average ego odom interval
  RCLCPP_INFO(get_logger(), "Calculating average ego odom interval");
  double total_interval = 0.0;
  for (size_t i = 1; i < rosbag_ego_odom_data_.size(); ++i) {
    const auto interval =
      (rosbag_ego_odom_data_[i].first - rosbag_ego_odom_data_[i - 1].first).seconds();
    total_interval += interval;
  }
  const double average_ego_odom_interval =
    total_interval / static_cast<double>(rosbag_ego_odom_data_.size() - 1);
  RCLCPP_INFO(get_logger(), "Average ego odom interval: %.3f seconds", average_ego_odom_interval);

  // start main timer callback
  RCLCPP_INFO(
    get_logger(), "Creating main timer with interval %.3f seconds", average_ego_odom_interval);
  timer_ = rclcpp::create_timer(
    this, get_clock(), std::chrono::duration<double>(average_ego_odom_interval),
    std::bind(&PerceptionReproducer::on_timer, this), callback_group_check_perception_);

  if (param_.publish_route) {
    publish_recorded_ego_pose(get_bag_start_time());
    // temporarily add a sleep because sometimes the route is not generated correctly without it.
    // Need to consider a proper solution.
    rclcpp::sleep_for(std::chrono::seconds(2));
    publish_goal_pose();
  }

  RCLCPP_INFO(get_logger(), "PerceptionReproducer initialization completed");
}

void PerceptionReproducer::on_initialpose(
  const geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr msg)
{
  (void)msg;
  cool_down_indices_.clear();

  last_sequenced_ego_pose_.reset();

  if (!last_published_timestamp_.has_value()) {
    const auto nearest_ego_odom_idx = find_nearest_ego_odom_index(msg->pose.pose);
    last_published_timestamp_ = rosbag_ego_odom_data_[nearest_ego_odom_idx].first;
  }

  RCLCPP_INFO(get_logger(), "Cool down indices and last sequenced pose cleared by /initialpose");
}

void PerceptionReproducer::on_timer()
{
  const auto timer_start = std::chrono::high_resolution_clock::now();
  const auto current_timestamp = this->get_clock()->now();

  // check if ego_odom is available
  const auto ego_odom_opt = get_latest_ego_odom();
  if (!ego_odom_opt.has_value()) {
    RCLCPP_WARN_THROTTLE(get_logger(), *get_clock(), 3000, "No ego odom found.");
    return;
  }

  const auto ego_odom = ego_odom_opt.value();
  const auto ego_pose = ego_odom.pose.pose;

  // calculate distance moved since last sequenced ego pose
  const double dist_moved =
    last_sequenced_ego_pose_.has_value()
      ? utils::calculate_distance_2d(ego_pose.position, last_sequenced_ego_pose_->position)
      : 999.0;

  // update the reproduce sequence if the distance moved is greater than the search radius
  if (dist_moved > ego_odom_search_radius_) {
    last_sequenced_ego_pose_ = ego_pose;

    // find nearby ego odom indices
    const auto nearest_ego_odom_idx = find_nearest_ego_odom_index(ego_pose);
    std::vector<geometry_msgs::msg::Pose> nearby_ego_odom_poses;

    const auto nearby_indices_initial =
      find_nearby_ego_odom_indices({ego_pose}, ego_odom_search_radius_);
    if (!nearby_indices_initial.empty()) {
      for (const auto idx : nearby_indices_initial) {
        nearby_ego_odom_poses.push_back(rosbag_ego_odom_data_[idx].second.pose.pose);
      }
    } else {
      nearby_ego_odom_poses.push_back(rosbag_ego_odom_data_[nearest_ego_odom_idx].second.pose.pose);
    }

    // find a list of ego odom around the nearest ego odom position
    auto ego_odom_indices =
      find_nearby_ego_odom_indices(nearby_ego_odom_poses, param_.search_radius);

    // remove indices that are in cool down list
    while (!cool_down_indices_.empty()) {
      const auto last_timestamp = ego_odom_id2last_published_timestamp_[cool_down_indices_.front()];
      if ((current_timestamp - last_timestamp).seconds() > param_.reproduce_cool_down) {
        cool_down_indices_.pop_front();
      } else {
        break;
      }
    }

    // update reproduce_sequence with data not in cool down list
    ego_odom_indices.erase(
      std::remove_if(
        ego_odom_indices.begin(), ego_odom_indices.end(),
        [this](const size_t idx) {
          return std::find(cool_down_indices_.begin(), cool_down_indices_.end(), idx) !=
                 cool_down_indices_.end();
        }),
      ego_odom_indices.end());

    std::sort(ego_odom_indices.begin(), ego_odom_indices.end());
    reproduce_sequence_indices_ =
      std::deque<size_t>(ego_odom_indices.begin(), ego_odom_indices.end());
  }

  if (param_.verbose) {
    std::string indices_str;
    for (size_t i = 0; i < std::min<size_t>(20, reproduce_sequence_indices_.size()); ++i) {
      indices_str += std::to_string(reproduce_sequence_indices_[i]) + " ";
    }
    RCLCPP_INFO(get_logger(), "reproduce_sequence_indices: %s", indices_str.c_str());
  }

  // determine whether to repeat previous message
  bool repeat_flag = reproduce_sequence_indices_.empty();

  // add constraint to avoid publishing too fast when speed gap exists
  if (!repeat_flag) {
    const double ego_speed = utils::calculate_speed_2d(ego_odom.twist.twist);

    const size_t ego_odom_idx = reproduce_sequence_indices_.front();
    const auto & ego_odom_msg = rosbag_ego_odom_data_[ego_odom_idx].second;
    const double ego_rosbag_speed = utils::calculate_speed_2d(ego_odom_msg.twist.twist);

    const double ego_rosbag_dist =
      utils::calculate_distance_2d(ego_pose.position, ego_odom_msg.pose.pose.position);

    repeat_flag = (ego_rosbag_speed > ego_speed * 2.0) && (ego_rosbag_speed > 3.0) &&
                  (ego_rosbag_dist > ego_odom_search_radius_);
  }

  // publish messages
  const auto bag_timestamp = [&]() -> std::optional<rclcpp::Time> {
    if (!repeat_flag) {
      const size_t ego_odom_idx = reproduce_sequence_indices_.front();
      reproduce_sequence_indices_.pop_front();

      const auto pose_timestamp = rosbag_ego_odom_data_[ego_odom_idx].first;

      // save the timestamp for potential repeat
      last_published_timestamp_ = pose_timestamp;

      // update cool down info
      ego_odom_id2last_published_timestamp_[ego_odom_idx] = current_timestamp;
      cool_down_indices_.push_back(ego_odom_idx);

      return pose_timestamp;
    }

    return last_published_timestamp_;
  }();

  if (bag_timestamp.has_value()) {
    publish_topics_at_timestamp(
      bag_timestamp.value(), current_timestamp, param_.noise && repeat_flag);
  } else {
    RCLCPP_WARN_THROTTLE(get_logger(), *get_clock(), 3000, "No valid bag timestamp to publish.");
  }

  const auto timer_end = std::chrono::high_resolution_clock::now();
  const auto total_time =
    std::chrono::duration_cast<std::chrono::microseconds>(timer_end - timer_start).count() / 1000.0;

  if (param_.verbose) {
    RCLCPP_INFO(get_logger(), "on_timer processing time: %.3f ms", total_time);
  }
}

size_t PerceptionReproducer::find_nearest_ego_odom_index(
  const geometry_msgs::msg::Pose & ego_pose) const
{
  const double target_x = ego_pose.position.x;
  const double target_y = ego_pose.position.y;

  double min_dist_squared = std::numeric_limits<double>::max();
  size_t nearest_idx = 0;

  for (size_t i = 0; i < rosbag_ego_odom_data_.size(); ++i) {
    const auto & odom_pos = rosbag_ego_odom_data_[i].second.pose.pose.position;
    const double dx = odom_pos.x - target_x;
    const double dy = odom_pos.y - target_y;
    const double dist_squared = dx * dx + dy * dy;

    if (dist_squared < min_dist_squared) {
      min_dist_squared = dist_squared;
      nearest_idx = i;
    }
  }

  return nearest_idx;
}

std::vector<size_t> PerceptionReproducer::find_nearby_ego_odom_indices(
  const std::vector<geometry_msgs::msg::Pose> & ego_poses, const double search_radius) const
{
  const double search_radius_squared = search_radius * search_radius;
  std::vector<size_t> nearby_indices;

  for (size_t i = 0; i < rosbag_ego_odom_data_.size(); ++i) {
    const auto & odom_pos = rosbag_ego_odom_data_[i].second.pose.pose.position;

    for (const auto & ego_pose : ego_poses) {
      const double dx = odom_pos.x - ego_pose.position.x;
      const double dy = odom_pos.y - ego_pose.position.y;
      const double dist_squared = dx * dx + dy * dy;

      if (dist_squared <= search_radius_squared) {
        nearby_indices.push_back(i);
        break;  // found within radius, no need to check other ego_poses
      }
    }
  }

  return nearby_indices;
}

}  // namespace autoware::planning_debug_tools
